### ISSUE TITLE
[raysgd] Invalid Device Ordinal in _set_cuda_device (Torch version 1.5)

### ISSUE THREAD

**Issue by DavidMChan** [ray-project/ray#9202]
<!--Please include [tune], [rllib], [autoscaler] etc. in the issue title if relevant-->

### What is the problem?

Ray can sometimes encounter a cuda runtime error (101): invalid device ordinal when setting devices - even when setting the string correctly to devices that exist on the system. This is similar to #5737 and #3772

I encountered this on the following machine:
OS: Ubuntu 18.04
Ray Version: 0.9.0.dev0 (master)
GPUs: 2x NVIDIA Titan (Maxwell)
CUDA version: 10.2
Torch version: 1.5

### Reproduction (REQUIRED)
This error can be encountered sporadically by launching any job with two workers, with 1 GPU per worker.

Note: This bug is flaky, and does not always occur. I can't seem to trigger it at will, suggesting that there may be a race condition when setting the device from multiple threads simultaneously (meaning that it may not be a ray bug, and might be something upstream).

Traceback:
```
Reading in data...                                                                                                                                                                                                     
Building configurations...                                                                                                                                                                                             
Configuring Training -- Iteration 0                                                                                                                                                                                    
Training -- Iteration 0                                                                                                                                                                                                
2020-06-29 19:02:15,938 INFO resource_spec.py:221 -- Starting Ray with 36.18 GiB memory available for workers and up to 18.09 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_
memory=<bytes>).                                                                                                                                                                                                       
2020-06-29 19:02:16,650 INFO services.py:1192 -- View the Ray dashboard at localhost:8265                                                                                                                              
2020-06-29 19:02:18,434 INFO distributed_torch_runner.py:349 -- Setting local CUDA device: 1                                                                                                                           
THCudaCheck FAIL file=/pytorch/torch/csrc/cuda/Module.cpp line=59 error=101 : invalid device ordinal                                                                                                                   
2020-06-29 19:02:18,438 ERROR distributed_torch_runner.py:353 -- Failed to set local CUDA device.                                                                                                                      
Traceback (most recent call last):                                                                                                                                                                                     
  File "active/driver.py", line 261, in <module>                                                                                                                                                                       
    main()                                                                                                                                                                                                             
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/click/core.py", line 829, in __call__                                                                                                        
    return self.main(*args, **kwargs)                                                                                                                                                                                  
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/click/core.py", line 782, in main                                                                                                            
    rv = self.invoke(ctx)                                                                                                                                                                                              
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/click/core.py", line 1066, in invoke                                                                                                         
    return ctx.invoke(self.callback, **ctx.params)                                                                                                                                                                     
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/click/core.py", line 610, in invoke                                                                                                          
    return callback(*args, **kwargs)                                                                                                                                                                                   
  File "active/driver.py", line 157, in main                                                                                                                                                                           
    results, checkpoint_path = train(config)                                                                                                                                                                           
  File "/home/david/Projects/pele/pele/run/train.py", line 85, in train                                                                                                                                                
    trainer = experiment.get_torch_trainer()                                                                                                                                                                           
  File "/home/david/Projects/pele/pele/core/experiment.py", line 228, in get_torch_trainer                                                                                                                             
    config=config_dict)                                                                                                                                                                                                
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/ray/util/sgd/torch/torch_trainer.py", line 276, in __init__                                                                                  
    self._start_workers(self.max_replicas)                                                                                                                                                                             
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/ray/util/sgd/torch/torch_trainer.py", line 336, in _start_workers                                                                            
    **params)                                                                                                                                                                                                          
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/ray/util/sgd/torch/distributed_torch_runner.py", line 307, in __init__                                                                       
    self._try_reserve_and_set_resources(num_cpus, num_gpus)                                                                                                                                                            
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/ray/util/sgd/torch/distributed_torch_runner.py", line 336, in _try_reserve_and_set_resources                                                 
    self._set_cuda_device(reserved_cuda_device)                                                                                                                                                                        
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/ray/util/sgd/torch/distributed_torch_runner.py", line 351, in _set_cuda_device                                                               
    torch.cuda.set_device(int(self.local_cuda_device))                                                                                                                                                                 
  File "/home/david/Apps/miniconda3/envs/pele/lib/python3.7/site-packages/torch/cuda/__init__.py", line 292, in set_device                                                                                             
    torch._C._cuda_setDevice(device)                                                                                                                                                                                   
RuntimeError: cuda runtime error (101) : invalid device ordinal at /pytorch/torch/csrc/cuda/Module.cpp:59 
```

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by richardliaw** [comment]
Thanks for opening this! Two questions:

1. Are you setting cuda visible devices?
2. if not, can you see if `torch.cuda.set_device(1)` or `torch.cuda.set_device(0)` or `torch.cuda.set_device("1")`  works on your interpreter?

**Comment by DavidMChan** [comment]
Thanks for following up - 
1. I am not using CUDA_VISIBLE_DEVICES
2. All of `torch.cuda.set_device(1)`, `torch.cuda.set_device(0)` and `torch.cuda.set_device('1')` work as expected in my interpreter.

**Comment by richardliaw** [comment]
ah hm I guess this might be a torch 1.5 thing - 
we set the cuda visible devices at runtime to mask GPUs. IIRC in torch 1.4 the setting of cuda visible devices at runtime isn't respected, while maybe torch 1.5 it is.

The temporary fix is to set `os.environ("CUDA_VISIBLE_DEVICES")="0"` right before TorchTrainer. 

I'll try to get to this this week. Let me know if the temporary fix works for you - sorry about the inconvenience!

**Comment by DavidMChan** [comment]
Thanks! I'll add that to my training scripts. It's not a major priority (since I can just rerun it a few times, and that seems to solve the problem), but it is a bit annoying.

**Comment by richardliaw** [comment]
Thanks! tip: just be careful not to set it before `ray.init`.

**Comment by DavidMChan** [comment]
So it seems like this doesn't consistently solve the issue. I believe that the issue is that ray sets CUDA_VISIBLE_DEVICES, however the local_device_string is always the same as the gpu number as in CUDA_VISIBLE_DEVICES (i.e. if the GPU visible is 3, then local_device_string is 3, where it should be 0, since it's the 0'th index when visible to torch). 

Additionally, to make matters worse  with NCCL 2.8, it looks like CUDA_VISIBLE_DEVICES is respected, meaning that when you get to DDP, all of the reserved devices have to be visible in CUDA_VISIBLE_DEVICES. (https://github.com/NVIDIA/nccl/issues/129)

Removing the set CUDA_VISIBLE_DEVICES from here: 
https://github.com/ray-project/ray/blob/f43d934817ae08119b38b59fb5dbcb68d2c8120d/python/ray/util/sgd/torch/distributed_torch_runner.py#L319
allows the script to continue correctly until failing in the DDP stage with a cuda memory illegal access error.

This doesn't seem to be an issue with Torch 1.4/1.5 (since I can reproduce it with both scenarios). It also seems to occur in cuda 10.0, 10.2 and 11.0

This may also be causing #9258, however I don't have enough information yet to debug it. 

@richardliaw - If you'd like to talk offline, I can give you access to a gcp instance where you can play with the code/errors.

**Comment by richardliaw** [comment]
btw, I think this actually should be fixed in #9260

**Comment by DavidMChan** [comment]
I can now reproduce the bug consistently with the script here:https://gist.github.com/DavidMChan/a3b88dbe16dacd368c9d3415acc174ae
The code is even worse if you move the ray.init()/ray.shutdown() inside the loop or increase the number of workers beyond 4. 

I'm not convinced that the fix #9260 fixes the problem entirely, since I still have the same issues in my larger code (however it reduces the number of issues caused). Adding `torch.cuda.is_initialized()` still encounters a race condition for the first thread initializing cuda.

I'm not certain that the fix in #9260 fixes all of the issues, especially when you have multiple trainers running in the same ray process. In fact, I think there may be deeper issues with multiple trainers than just these bugs (but I don't know, since my development loop is pretty slow).

**Comment by richardliaw** [comment]
hm got it; I think to really fix this issue we'll have to have an option to make *all* workers distributed.

**Comment by richardliaw** [comment]
This is actually fixed in #10539!

**Comment by richardrl** [comment]
@richardliaw I am on 1.2.0dev0 and still experience this issue with invalid device ordinal. I am giving each trial 2 GPUs on a 4 GPU system, and using this code to set the GPUs:

```
    for idx, (key, model) in enumerate(models_container_dict.items()):
        selected_gpu = gpu_device_pool[idx % len(gpu_device_pool)]
        print("Selected GPU")
        print(selected_gpu)
        model.to(torch.device(f"cuda:{selected_gpu}"))
```
